### ISSUE TITLE
DOC: Style guide could indicate formatters / pre-commit hook

### ISSUE THREAD

**Issue by tgross35** [numpy/numpy#21424]
### Issue with current documentation:

I'm working through my first issue with the project, and struggling to find information on whether there is a recommended formatter. The source currently seems to use a mix of optional styles (single vs. double quotes, backslash vs. brackets, formatting of non-python files, etc.) and I think that some more specific in the docs, possibly with recommended formatters and/or formatter config files, would help to simplify things a bit.

The only reference to style that I found was here: https://numpydoc.readthedocs.io/en/latest/format.html

### Idea or request for content:

Plenty of projects use Black to format code, and I don't think that would be a bad idea here. If there are some custom style options that Numpy prefers to stick with, then something like Autopep8 or YAPF would probably be configurable to work.

A simpler option would be to add a precommit hook that handles formatting, just to be able to easier enforce consistency. Even if the hook isn't enforced in the pipeline, it would at least help give devs some direction when working locally. At a minimum, the following hooks would be quite useful:
- Black (with line length 79 to match pycodestyle)
- flake8, recommended in the Numpy style guide
- trailing-whitespace
- end-of-file-fixer
- clang-format, to implement the config from here https://github.com/numpy/numpy/pull/19754

Optionally, some other nice to have hooks would be:
- isort
- pretty-format-ini
- pretty-format-toml
- pretty-format-yaml
- check-toml
- check-yaml
- fix-byte-order-marker
- mixed-line-ending

If there is interest in this, I would be happy to submit a PR. Of course, it is possible that this has been discussed before and there's a reason it doesn't exist. I couldn't find anything in the issues list, but I figure there must be a reason why linter.py is used instead of a precommit hook.

**Comment by mattip** [comment]
There is a CI linter job that runs a script `tools/linter.py`. This is a comprimise solution to only check lines that have changed. The PR to add this was https://github.com/numpy/numpy/pull/18423 which was transfered from scipy https://github.com/scipy/scipy/pull/11423 and based on a mailing list discussion https://mail.python.org/pipermail/scipy-dev/2020-January/023927.html.

**Comment by tgross35** [comment]
Thank you for the links, makes sense about the incremental changes.

If there is a goal of eventually having some sort of format, [darker](https://pypi.org/project/darker/) exists which runs black & isort only on changed lines, and can be a precommit hook or an editor plugin. There's also a way to do the same for a hook with clang-format, which would allow for your comment in https://github.com/numpy/numpy/pull/19754#issuecomment-910312569

If it's not a "hard no" to have some sort of optional incremental formatter, then I'll come up with a sample config file to at least test out

**Comment by DimitriPapadopoulos** [comment]
Some of these suggestions have been applied incrementally:
* use `ruff` as a linter,
* enforce rulesets such as [pycodestyle (E, W)](https://docs.astral.sh/ruff/rules/#pycodestyle-e-w) or [pyupgrade (UP)](https://docs.astral.sh/ruff/rules/#pyupgrade-up).

Not done yet:
* enforce [isort](https://docs.astral.sh/ruff/rules/#isort-i) rules,
* enforce [flake8-bugbear (B)](https://docs.astral.sh/ruff/rules/#flake8-bugbear-b) rules as suggested by [Repo-Review](https://learn.scientific-python.org/development/guides/repo-review/?repo=numpy%2Fnumpy&ref=main&refType=branch),
* use pre-commit hooks (?).

It remains unclear if there is interest in pre-commit hooks, and whether they should be enforced by [pre-cmomit.ci](https://pre-commit.ci/) or the official GitHub Action https://github.com/pre-commit/action.

**Comment by jorenham** [comment]
I would recommend against using pre-commit actually. See https://github.com/numpy/numtype/issues/388#issuecomment-2746333410 and https://github.com/scientific-python/cookie/issues/577

**Comment by DimitriPapadopoulos** [comment]
I'm not a huge fan of pre-commit either, for different reasons, and I have to agree the maintainer is.. difficult.

**Comment by tgross35** [comment]
Yeah, I think the situation has improved here and the reasons to avoid pre-commit have grown, so this can be closed.
